File: solutions/clojure/robot-simulator/1/src/robot_simulator.clj
(ns robot-simulator)

(defn- turn [direction instruction]
  (let [transitions {:north {"R" :east,  "L" :west}
                     :south {"R" :west,  "L" :east}
                     :east  {"R" :south, "L" :north}
                     :west  {"R" :north, "L" :south}}]
    (get-in transitions [direction instruction])))

(defn- process-instruction [instruction robot-state]
  (let [bearing (if (or (= instruction "R") (= instruction "L"))
                    (turn (:bearing robot-state) instruction)
                    (:bearing robot-state)
                  )
        coordinates (:coordinates robot-state)
        updated-coordinates (if (= instruction "A")
                              (case bearing
                                  :north (update coordinates :y inc)
                                  :south (update coordinates :y dec)
                                  :east (update coordinates :x inc)
                                  :west (update coordinates :x dec)
                                )
                              coordinates)]
      {:bearing bearing :coordinates updated-coordinates}
    )
  )

(defn robot
  "Creates a robot at the given coordinates, facing the given direction."
  [coordinates direction]
    {:bearing direction :coordinates coordinates}
  )

(defn simulate
  "Simulates the robot's movements based on the given instructions
  and updates its state."
  [instructions robot-state]
    (reduce (fn [state instruction]
              (process-instruction (str instruction) state))
            robot-state
            instructions)
  )

File: solutions/python/robot-simulator/1/robot_simulator.py
EAST = (1,0) 
NORTH = (0,1)
WEST = (-1,0)
SOUTH = (0,-1)

class Robot:
    def __init__(self, direction=NORTH, x_pos=0, y_pos=0):
        self.coordinates = (x_pos, y_pos)
        self.direction = direction

    def do_move(self, movement):
        if movement == "R":
            self.direction = (self.direction[1], -self.direction[0])
        elif movement == "L":
            self.direction = (-self.direction[1], self.direction[0])
        elif movement == "A":
            self.coordinates = (self.coordinates[0] + self.direction[0], self.coordinates[1] + self.direction[1])

    def move(self, movements):
        for movement in movements:
            self.do_move(movement)
